File: packages/api-client/src/schemas.ts
import { Schema } from "effect";

/**
 * Plain `Schema.Struct`s, not `Schema.Class`es: decoded payloads cross the
 * TanStack Start SSR boundary, and Seroval only serializes plain objects.
 */

export const HealthResponse = Schema.Struct({
  status: Schema.String,
  offline: Schema.Boolean,
  fixtures_loaded: Schema.Int,
});
export type HealthResponse = typeof HealthResponse.Type;

export const RunState = Schema.Literals([
  "queued",
  "running",
  "needs_more_data",
  "awaiting_review",
  "complete",
  "failed",
]);
export type RunState = typeof RunState.Type;

export const Verdict = Schema.Literals(["compliant", "high_risk", "ambiguous"]);
export type Verdict = typeof Verdict.Type;

export const RiskLevel = Schema.Literals(["negligible", "non_negligible"]);
export type RiskLevel = typeof RiskLevel.Type;

export const FindingCode = Schema.Literals([
  "deforestation_loss",
  "fire_cluster",
  "legal_permit",
  "area_mismatch",
  "certification",
]);
export type FindingCode = typeof FindingCode.Type;

export const FindingLevel = Schema.Literals(["clear", "borderline", "flag"]);
export type FindingLevel = typeof FindingLevel.Type;

export const EvidenceSource = Schema.Literals([
  "global_forest_watch",
  "nasa_firms",
  "legality_dataset",
  "concession_dataset",
  "consignment_dataset",
  "operator_dataset",
  "deterministic_rubric",
]);
export type EvidenceSource = typeof EvidenceSource.Type;

export const TraceKind = Schema.Literals([
  "run",
  "agent",
  "tool",
  "verifier",
  "writer",
  "review",
  "state",
]);
export type TraceKind = typeof TraceKind.Type;

export const ModelMode = Schema.Literals(["scripted", "bedrock"]);
export type ModelMode = typeof ModelMode.Type;

export const Finding = Schema.Struct({
  code: FindingCode,
  level: FindingLevel,
  points: Schema.Int,
  detail: Schema.String,
  metrics: Schema.Record(
    Schema.String,
    Schema.Union([Schema.Number, Schema.String]),
  ),
  evidence_ids: Schema.Array(Schema.String),
  mitigating: Schema.Boolean,
});
export type Finding = typeof Finding.Type;

export const Assessment = Schema.Struct({
  record_id: Schema.NullOr(Schema.String),
  supplier_id: Schema.String,
  polygon_id: Schema.String,
  rubric_version: Schema.String,
  score: Schema.Int,
  verdict: Verdict,
  risk_level: RiskLevel,
  findings: Schema.Array(Finding),
  citations: Schema.Record(Schema.String, Schema.Array(Schema.String)),
  disclosures: Schema.Array(Schema.String),
  data_gaps: Schema.Array(Schema.String),
});
export type Assessment = typeof Assessment.Type;

export const VerificationChallenge = Schema.Struct({
  kind: Schema.String,
  severity: Schema.Literals(["info", "warning", "error"]),
  detail: Schema.String,
  source: Schema.String,
  evidence_ids: Schema.Array(Schema.String),
  expected: Schema.NullOr(Schema.String),
  observed: Schema.NullOr(Schema.String),
});
export type VerificationChallenge = typeof VerificationChallenge.Type;

export const VerificationReport = Schema.Struct({
  accepted: Schema.Boolean,
  checked_claims: Schema.Array(Schema.String),
  challenges: Schema.Array(VerificationChallenge),
  notes: Schema.Array(Schema.String),
  llm_reviewed: Schema.Boolean,
  model_id: Schema.NullOr(Schema.String),
});
export type VerificationReport = typeof VerificationReport.Type;

export const TraceStep = Schema.Struct({
  step_id: Schema.String,
  kind: TraceKind,
  name: Schema.String,
  detail: Schema.String,
  at: Schema.DateTimeUtcFromString,
  payload: Schema.Record(Schema.String, Schema.Unknown),
});
export type TraceStep = typeof TraceStep.Type;

export const ReviewOut = Schema.Struct({
  decision: Schema.Literals(["approve", "override"]),
  reviewer: Schema.String,
  note: Schema.String,
  reviewed_at: Schema.NullOr(Schema.DateTimeUtcFromString),
});
export type ReviewOut = typeof ReviewOut.Type;

export const DdsMeta = Schema.Struct({
  released: Schema.Boolean,
  schema_version: Schema.String,
  created_at: Schema.DateTimeUtcFromString,
});
export type DdsMeta = typeof DdsMeta.Type;

const runSummaryFields = {
  run_id: Schema.String,
  kind: Schema.String,
  state: RunState,
  record_id: Schema.NullOr(Schema.String),
  supplier_id: Schema.NullOr(Schema.String),
  polygon_id: Schema.NullOr(Schema.String),
  parent_run_id: Schema.NullOr(Schema.String),
  model: Schema.String,
  summary: Schema.String,
  error: Schema.NullOr(Schema.String),
  verdict: Schema.NullOr(Verdict),
  score: Schema.NullOr(Schema.Int),
  expected_archetype: Schema.NullOr(Schema.String),
  step_count: Schema.Int,
  started_at: Schema.NullOr(Schema.DateTimeUtcFromString),
  finished_at: Schema.NullOr(Schema.DateTimeUtcFromString),
  elapsed_seconds: Schema.NullOr(Schema.Number),
  created_at: Schema.DateTimeUtcFromString,
  metrics: Schema.Record(Schema.String, Schema.Unknown),
} as const;

export const RunSummary = Schema.Struct(runSummaryFields);
export type RunSummary = typeof RunSummary.Type;

export const VerdictOut = Schema.Struct({
  record_id: Schema.String,
  supplier_id: Schema.String,
  polygon_id: Schema.String,
  run_state: RunState,
  expected_archetype: Schema.String,
  expected_signal: Schema.NullOr(Schema.String),
  expected_ambiguity: Schema.NullOr(Schema.String),
  fingerprint: Schema.String,
  pending: Schema.Boolean,
  assessment: Assessment,
});
export type VerdictOut = typeof VerdictOut.Type;

export const RunDetail = Schema.Struct({
  ...runSummaryFields,
  disclosures: Schema.Array(Schema.String),
  verification: Schema.NullOr(VerificationReport),
  assessment: Schema.NullOr(VerdictOut),
  pending_assessment: Schema.NullOr(VerdictOut),
  review: Schema.NullOr(ReviewOut),
  steps: Schema.Array(TraceStep),
  dds: Schema.NullOr(DdsMeta),
});
export type RunDetail = typeof RunDetail.Type;

export const BatchSummary = Schema.Struct({
  run_id: Schema.String,
  state: RunState,
  record_count: Schema.Int,
  states: Schema.Record(Schema.String, Schema.Int),
  verdict_breakdown: Schema.Record(Schema.String, Schema.Int),
  expected_breakdown: Schema.Record(Schema.String, Schema.Int),
  wall_clock_seconds: Schema.NullOr(Schema.Number),
  average_seconds_per_record: Schema.NullOr(Schema.Number),
  started_at: Schema.NullOr(Schema.DateTimeUtcFromString),
  finished_at: Schema.NullOr(Schema.DateTimeUtcFromString),
  metrics: Schema.Record(Schema.String, Schema.Unknown),
});
export type BatchSummary = typeof BatchSummary.Type;

export const ParcelOut = Schema.Struct({
  polygon_id: Schema.String,
  supplier_id: Schema.String,
  label: Schema.String,
  region: Schema.String,
  province: Schema.String,
  area_ha: Schema.Number,
  centroid_lat: Schema.Number,
  centroid_lon: Schema.Number,
  geometry: Schema.Record(Schema.String, Schema.Unknown),
  archetype: Schema.String,
  scenario: Schema.NullOr(Schema.String),
  is_demo: Schema.Boolean,
});
export type ParcelOut = typeof ParcelOut.Type;

const supplierFields = {
  supplier_id: Schema.String,
  legal_name: Schema.String,
  trading_name: Schema.String,
  group: Schema.String,
  nib: Schema.String,
  npwp: Schema.String,
  hgu_number: Schema.NullOr(Schema.String),
  pbp_number: Schema.NullOr(Schema.String),
  permit_status: Schema.String,
  concession_area_ha: Schema.Number,
  province: Schema.String,
  kabupaten: Schema.String,
  beneficial_owners: Schema.Array(Schema.String),
  certifications: Schema.Array(Schema.String),
  sanctions: Schema.Array(Schema.String),
  synthetic: Schema.Boolean,
  disclosure: Schema.String,
} as const;

export const SupplierOut = Schema.Struct(supplierFields);
export type SupplierOut = typeof SupplierOut.Type;

export const SupplierDetail = Schema.Struct({
  ...supplierFields,
  parcels: Schema.Array(ParcelOut),
});
export type SupplierDetail = typeof SupplierDetail.Type;

export const EvidenceEntry = Schema.Struct({
  schema_version: Schema.Int,
  evidence_id: Schema.String,
  claim: Schema.String,
  value: Schema.Unknown,
  unit: Schema.NullOr(Schema.String),
  source: EvidenceSource,
  artifact: Schema.Record(Schema.String, Schema.Unknown),
  retrieved_at: Schema.NullOr(Schema.DateTimeUtcFromString),
  cached: Schema.Boolean,
  synthetic: Schema.Boolean,
  disclosure: Schema.NullOr(Schema.String),
});
export type EvidenceEntry = typeof EvidenceEntry.Type;

export const SapVendorOut = Schema.Struct({
  vendor_id: Schema.String,
  status: Schema.String,
  purchasing_block: Schema.Boolean,
});
export type SapVendorOut = typeof SapVendorOut.Type;

export const RunCreateRequest = Schema.Struct({
  record_id: Schema.optional(Schema.String),
  scenario: Schema.optional(Schema.String),
  model: Schema.optional(ModelMode),
  refresh: Schema.optional(Schema.Boolean),
});
export type RunCreateRequest = typeof RunCreateRequest.Type;

export const DecisionRequest = Schema.Struct({
  decision: Schema.Literals(["approve", "override"]),
  reviewer: Schema.optional(Schema.String),
  note: Schema.optional(Schema.String),
});
export type DecisionRequest = typeof DecisionRequest.Type;

export const BatchCreateRequest = Schema.Struct({
  size: Schema.optional(Schema.Int),
  refresh: Schema.optional(Schema.Boolean),
});
export type BatchCreateRequest = typeof BatchCreateRequest.Type;

export const SapStatusRequest = Schema.Struct({ status: Schema.String });
export type SapStatusRequest = typeof SapStatusRequest.Type;

export const SapBlockRequest = Schema.Struct({ blocked: Schema.Boolean });
export type SapBlockRequest = typeof SapBlockRequest.Type;

const SnapshotData = Schema.Struct({
  run_id: Schema.String,
  kind: Schema.String,
  state: RunState,
  record_id: Schema.NullOr(Schema.String),
  parent_run_id: Schema.NullOr(Schema.String),
  verdict: Schema.NullOr(Verdict),
  score: Schema.NullOr(Schema.Int),
  dds_released: Schema.Boolean,
});

const StateData = Schema.Struct({
  run_id: Schema.String,
  state: RunState,
});

const ProgressData = Schema.Struct({
  run_id: Schema.String,
  total: Schema.Int,
  completed: Schema.Int,
  failed: Schema.Int,
  awaiting_review: Schema.Int,
  verdict_breakdown: Schema.Record(Schema.String, Schema.Int),
});

export const RunEvent = Schema.Union([
  Schema.Struct({ event: Schema.Literal("snapshot"), data: SnapshotData }),
  Schema.Struct({ event: Schema.Literal("step"), data: TraceStep }),
  Schema.Struct({ event: Schema.Literal("state"), data: StateData }),
  Schema.Struct({ event: Schema.Literal("done"), data: StateData }),
  Schema.Struct({ event: Schema.Literal("progress"), data: ProgressData }),
]);
export type RunEvent = typeof RunEvent.Type;

File: apps/api/src/terrasentry_api/schemas.py
"""Request/response models for the public API.

These are the OpenAPI contract: routers build them from the audit rows and the
generated TypeScript client mirrors them. Nested domain payloads (assessment,
verification, trace steps, evidence) come straight from the deterministic core
so the API cannot drift from the engine.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field, model_validator
from terrasentry_core.agents.verification import VerificationReport
from terrasentry_core.domain.enums import RunState, Verdict
from terrasentry_core.domain.models import Assessment, Finding
from terrasentry_core.tools.trace import TraceStep
from terrasentry_integrations.sap import VendorStatus

from terrasentry_api.models import DdsDocument as DdsRow
from terrasentry_api.models import Parcel as ParcelRow
from terrasentry_api.models import Run, RunVerdict
from terrasentry_api.models import SapAction as SapActionRow
from terrasentry_api.models import Supplier as SupplierRow
from terrasentry_api.sap_actions import disclosure_for_mode

ModelMode = Literal["scripted", "bedrock"]


def _opt_float(value: Any) -> float | None:
    return float(value) if isinstance(value, (int, float)) else None


class HealthResponse(BaseModel):
    """Process health plus the offline-rehearsal and SAP-mode state the cockpit badges."""

    status: str
    offline: bool = False
    fixtures_loaded: int = 0
    sap_mode: str = "stub"
    sap_real: bool = False


class RunCreate(BaseModel):
    """Start one scenario run for a seed record or a named demo scenario."""

    record_id: str | None = None
    scenario: str | None = None
    model: ModelMode | None = None
    refresh: bool = False

    @model_validator(mode="after")
    def _exactly_one_target(self) -> RunCreate:
        if bool(self.record_id) == bool(self.scenario):
            raise ValueError("provide exactly one of record_id or scenario")
        return self


class DecisionCreate(BaseModel):
    """Record the human decision on an ``awaiting_review`` run."""

    decision: Literal["approve", "override"]
    reviewer: str = "human"
    note: str = ""


class BatchCreate(BaseModel):
    """Queue a deterministic batch run over the seed records."""

    size: int = Field(default=50, ge=1, le=50)
    refresh: bool = False


class ReviewOut(BaseModel):
    decision: str
    reviewer: str
    note: str = ""
    reviewed_at: datetime | None = None


class VerdictOut(BaseModel):
    """The deterministic assessment of one record, pending or released."""

    record_id: str
    supplier_id: str
    polygon_id: str
    run_state: RunState
    expected_archetype: str
    expected_signal: str | None = None
    expected_ambiguity: str | None = None
    fingerprint: str
    pending: bool
    assessment: Assessment

    @classmethod
    def from_row(cls, row: RunVerdict) -> VerdictOut:
        return cls(
            record_id=row.record_id,
            supplier_id=row.supplier_id,
            polygon_id=row.polygon_id,
            run_state=RunState.AWAITING_REVIEW if row.pending else RunState.COMPLETE,
            expected_archetype=row.expected_archetype,
            expected_signal=row.expected_signal,
            expected_ambiguity=row.expected_ambiguity,
            fingerprint=row.fingerprint,
            pending=row.pending,
            assessment=Assessment(
                record_id=row.record_id,
                supplier_id=row.supplier_id,
                polygon_id=row.polygon_id,
                rubric_version=row.rubric_version,
                score=row.score,
                verdict=row.verdict,
                risk_level=row.risk_level,
                findings=[Finding.model_validate(finding) for finding in row.findings],
                citations=row.citations,
                disclosures=row.disclosures,
                data_gaps=row.data_gaps,
            ),
        )


class DdsMeta(BaseModel):
    """Presence and release state of the DDS; the document lives at /dds/{run_id}."""

    released: bool
    schema_version: str
    created_at: datetime

    @classmethod
    def from_row(cls, row: DdsRow) -> DdsMeta:
        return cls(
            released=row.released,
            schema_version=str(row.document.get("schemaVersion", "")),
            created_at=row.created_at,
        )


class RunSummary(BaseModel):
    """One run row as it appears in list views and the cockpit table."""

    run_id: str
    kind: str
    state: RunState
    record_id: str | None = None
    supplier_id: str | None = None
    polygon_id: str | None = None
    parent_run_id: str | None = None
    model: str
    summary: str = ""
    error: str | None = None
    verdict: Verdict | None = None
    score: int | None = None
    expected_archetype: str | None = None
    step_count: int = 0
    started_at: datetime | None = None
    finished_at: datetime | None = None
    elapsed_seconds: float | None = None
    created_at: datetime
    metrics: dict[str, Any] = Field(default_factory=dict)

    @classmethod
    def from_row(
        cls,
        run: Run,
        *,
        verdict: RunVerdict | None = None,
        step_count: int = 0,
    ) -> RunSummary:
        return cls(
            run_id=run.id,
            kind=run.kind,
            state=run.state,
            record_id=run.record_id,
            supplier_id=run.supplier_id,
            polygon_id=run.polygon_id,
            parent_run_id=run.parent_run_id,
            model=run.model,
            summary=run.summary,
            error=run.error,
            verdict=verdict.verdict if verdict is not None else None,
            score=verdict.score if verdict is not None else None,
            expected_archetype=verdict.expected_archetype if verdict is not None else None,
            step_count=step_count,
            started_at=run.started_at,
            finished_at=run.finished_at,
            elapsed_seconds=run.elapsed_seconds,
            created_at=run.created_at,
            metrics=run.metrics,
        )


class RunDetail(RunSummary):
    """Full dossier: trace, verification, assessment, review, SAP action, and DDS metadata."""

    disclosures: list[str] = Field(default_factory=list)
    verification: VerificationReport | None = None
    assessment: VerdictOut | None = None
    pending_assessment: VerdictOut | None = None
    review: ReviewOut | None = None
    sap_action: SapActionOut | None = None
    steps: list[TraceStep] = Field(default_factory=list)
    dds: DdsMeta | None = None

    @classmethod
    def from_parts(
        cls,
        run: Run,
        *,
        verdict: RunVerdict | None,
        dds: DdsRow | None,
        steps: list[TraceStep],
        sap_action: SapActionRow | None = None,
    ) -> RunDetail:
        summary = RunSummary.from_row(run, verdict=verdict, step_count=len(steps))
        review = None
        if run.review_decision is not None:
            review = ReviewOut(
                decision=run.review_decision,
                reviewer=run.review_reviewer or "human",
                note=run.review_note or "",
                reviewed_at=run.reviewed_at,
            )
        verdict_out = VerdictOut.from_row(verdict) if verdict is not None else None
        return cls(
            **summary.model_dump(),
            disclosures=run.disclosures,
            verification=(VerificationReport.model_validate(run.verification) if run.verification else None),
            assessment=verdict_out if verdict_out is not None and not verdict_out.pending else None,
            pending_assessment=(verdict_out if verdict_out is not None and verdict_out.pending else None),
            review=review,
            sap_action=SapActionOut.from_row(sap_action) if sap_action is not None else None,
            steps=steps,
            dds=DdsMeta.from_row(dds) if dds is not None else None,
        )


class BatchSummary(BaseModel):
    """Aggregate throughput view over one batch parent run."""

    run_id: str
    state: RunState
    record_count: int
    states: dict[str, int] = Field(default_factory=dict)
    verdict_breakdown: dict[str, int] = Field(default_factory=dict)
    expected_breakdown: dict[str, int] = Field(default_factory=dict)
    wall_clock_seconds: float | None = None
    average_seconds_per_record: float | None = None
    median_seconds_per_record: float | None = None
    p95_seconds_per_record: float | None = None
    throughput_records_per_second: float | None = None
    total_record_seconds: float | None = None
    cache_stats: dict[str, int] | None = None
    confusion: dict[str, dict[str, int]] = Field(default_factory=dict)
    batch_concurrency: int | None = None
    sap_actions: dict[str, int] = Field(default_factory=dict)
    started_at: datetime | None = None
    finished_at: datetime | None = None
    metrics: dict[str, Any] = Field(default_factory=dict)

    @classmethod
    def from_run(
        cls,
        run: Run,
        *,
        states: dict[str, int],
        breakdown: dict[str, int],
        average_seconds: float,
    ) -> BatchSummary:
        metrics = run.metrics or {}
        confusion = metrics.get("confusion")
        cache_stats = metrics.get("cache_stats")
        concurrency = metrics.get("batch_concurrency")
        sap_actions = metrics.get("sap_actions")
        return cls(
            run_id=run.id,
            state=run.state,
            record_count=int(metrics.get("record_count", 0)) or sum(states.values()),
            states=states,
            verdict_breakdown=breakdown,
            expected_breakdown=metrics.get("expected_breakdown", {}),
            wall_clock_seconds=run.elapsed_seconds,
            average_seconds_per_record=average_seconds,
            median_seconds_per_record=_opt_float(metrics.get("median_seconds_per_record")),
            p95_seconds_per_record=_opt_float(metrics.get("p95_seconds_per_record")),
            throughput_records_per_second=_opt_float(metrics.get("throughput_records_per_second")),
            total_record_seconds=_opt_float(metrics.get("total_record_seconds")),
            cache_stats=cache_stats if isinstance(cache_stats, dict) else None,
            confusion=confusion if isinstance(confusion, dict) else {},
            batch_concurrency=int(concurrency) if isinstance(concurrency, (int, float)) else None,
            sap_actions=sap_actions if isinstance(sap_actions, dict) else {},
            started_at=run.started_at,
            finished_at=run.finished_at,
            metrics=metrics,
        )


class ParcelOut(BaseModel):
    polygon_id: str
    supplier_id: str
    label: str
    region: str
    province: str
    area_ha: float
    centroid_lat: float
    centroid_lon: float
    geometry: dict[str, Any]
    archetype: str
    scenario: str | None = None
    is_demo: bool

    @classmethod
    def from_row(cls, row: ParcelRow) -> ParcelOut:
        return cls(
            polygon_id=row.polygon_id,
            supplier_id=row.supplier_id,
            label=row.label,
            region=row.region,
            province=row.province,
            area_ha=row.area_ha,
            centroid_lat=row.centroid_lat,
            centroid_lon=row.centroid_lon,
            geometry=row.geometry,
            archetype=row.archetype,
            scenario=row.scenario,
            is_demo=row.is_demo,
        )


class SupplierOut(BaseModel):
    supplier_id: str
    legal_name: str
    trading_name: str
    group: str
    nib: str
    npwp: str
    hgu_number: str | None = None
    pbp_number: str | None = None
    permit_status: str
    concession_area_ha: float
    province: str
    kabupaten: str
    beneficial_owners: list[str] = Field(default_factory=list)
    certifications: list[str] = Field(default_factory=list)
    sanctions: list[str] = Field(default_factory=list)
    synthetic: bool
    disclosure: str = ""

    @classmethod
    def from_row(cls, row: SupplierRow) -> SupplierOut:
        return cls(
            supplier_id=row.supplier_id,
            legal_name=row.legal_name,
            trading_name=row.trading_name,
            group=row.group,
            nib=row.nib,
            npwp=row.npwp,
            hgu_number=row.hgu_number,
            pbp_number=row.pbp_number,
            permit_status=row.permit_status,
            concession_area_ha=row.concession_area_ha,
            province=row.province,
            kabupaten=row.kabupaten,
            beneficial_owners=row.beneficial_owners,
            certifications=row.certifications,
            sanctions=row.sanctions,
            synthetic=row.synthetic,
            disclosure=row.disclosure,
        )


class SapActionOut(BaseModel):
    """One recorded ERP action from the M7 closed loop."""

    run_id: str
    supplier_id: str
    vendor_id: str
    mode: str
    status: str
    purchasing_block: bool
    real: bool
    external_reference: str | None = None
    error: str | None = None
    performed_at: datetime

    @classmethod
    def from_row(cls, row: SapActionRow) -> SapActionOut:
        return cls(
            run_id=row.run_id,
            supplier_id=row.supplier_id,
            vendor_id=row.vendor_id,
            mode=row.mode,
            status=row.status,
            purchasing_block=row.purchasing_block,
            real=row.real,
            external_reference=row.external_reference,
            error=row.error,
            performed_at=row.performed_at,
        )


class SupplierSapOut(BaseModel):
    """Current ERP state for one supplier plus its most recent action, if any."""

    vendor_id: str
    status: str
    purchasing_block: bool
    mode: str
    real: bool
    disclosure: str
    last_action: SapActionOut | None = None

    @classmethod
    def from_parts(
        cls,
        supplier_id: str,
        *,
        current: VendorStatus | None,
        last_action: SapActionRow | None,
        mode: str,
        real: bool,
    ) -> SupplierSapOut:
        vendor = current or VendorStatus(vendor_id=supplier_id)
        return cls(
            vendor_id=vendor.vendor_id,
            status=vendor.status,
            purchasing_block=vendor.purchasing_block,
            mode=mode,
            real=real,
            disclosure=disclosure_for_mode(mode),
            last_action=SapActionOut.from_row(last_action) if last_action is not None else None,
        )


class SupplierDetail(SupplierOut):
    parcels: list[ParcelOut] = Field(default_factory=list)
    sap: SupplierSapOut | None = None


class SapVendorOut(BaseModel):
    vendor_id: str
    status: str
    purchasing_block: bool

    @classmethod
    def from_vendor(cls, vendor: VendorStatus) -> SapVendorOut:
        return cls(
            vendor_id=vendor.vendor_id,
            status=vendor.status,
            purchasing_block=vendor.purchasing_block,
        )


class SapStatusIn(BaseModel):
    status: str = Field(min_length=1, max_length=32)


class SapBlockIn(BaseModel):
    blocked: bool


__all__ = [
    "BatchCreate",
    "BatchSummary",
    "DdsMeta",
    "DecisionCreate",
    "ParcelOut",
    "ReviewOut",
    "RunCreate",
    "RunDetail",
    "RunSummary",
    "SapActionOut",
    "SapBlockIn",
    "SapStatusIn",
    "SapVendorOut",
    "SupplierDetail",
    "SupplierOut",
    "SupplierSapOut",
    "VerdictOut",
]
